tracing-core/src/callsite.rs

```diff
@@ -169,3 +177,169 @@ impl Hash for Identifier {
         (self.0 as *const dyn Callsite).hash(state)
     }
 }
+
+/// An intrusive atomic push only linked-list.
+struct LinkedList {
+    head: AtomicPtr<Registration>,
+}
+
+unsafe impl Send for LinkedList {}
+unsafe impl Sync for LinkedList {}
+
+/// Represents a registration within the callsite cache.
+///
+/// This type enables the intrusive aspect of the callsite cache by
+/// avoiding allocations and letting the callsite's themselves store
+/// the next callsite in the cache.
+pub struct Registration {
+    #[doc(hidden)]
+    pub callsite: &'static dyn Callsite,
+    #[doc(hidden)]
+    pub next: AtomicPtr<Registration>,
```

Comment:
is there a reason this needs to have `pub` fields? I know we can't have a `const fn` constructor that takes a `&'static dyn Callsite`, but i think there's a workaround for this.

i think we can change the type to
```rust
pub struct Registration<T = &'static dyn Callsite> {
    callsite: T,
    next: AtomicPtr<Registration<T>>,
}

impl<T> Registration<T> {
    pub const fn new(callsite: T) -> Self {
        Self {
            callsite,
            next: AtomicPtr::new(ptr::null_mut()),
        }
    }
}
```
and leave everything else as-is. although it will be technically _possible_ to construct a `Registration` with any type, the `Callsite` trait will still require a `Registration` contain a `&'static dyn Callsite`, and (because of the default type param), the `Registration` type will not appear to be generic in the docs except for the actual constructor.

Reply:
specifically, i would really prefer it if user code was not able to see or touch the pointer...